### ISSUE TITLE
Memory leaks due to unmounting components with alive promises

### ISSUE THREAD

**Issue by richard-ejem** [reactjs/reactjs.org#654]
Hi,
I just came to [isMounted is an Antipattern article](https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html) . I'm afraid that the solution using a cancelable promise you suggest here does not solve the potential memory leak, but hides it basically the same way as using `isMounted` would do.

Why?
The problem is that when a component starts some asynchronous operation (i.e. XHR) and passes a callback which should update component's state to a Promise or Promise-like interface, the callback holds a reference to the component so until the callback object exists, the component cannot be fully removed from memory.
When the cancelable promise you introduce here is canceled, no references are broken, just a boolean flag is set. Therefore the component stays in memory along with the promise.
In other words, it would be easy to un-cancel such promise by setting `hasCancelled_` back to `false`, therefore the reference to the object have to still exist. The only thing this cancellable promise does (not) is skipping the call to `resolve(val)`, but from the memory management view it is the same approach as asking for `isMounted` inside the component.

My opinion is that the only proper solution would be to abort async operations inside the promise (clear timeouts, abort XHRs) and `delete` the promise itself.

**Comment by Alex1100** [comment]
@richard-ejem, Wouldn't something like this work. I kept running into an Error within an app I was working on and I basically came up with this to avoid having the last cb crash and return an Error.

```tsx
interface CancellablePromise {
  promise: Promise<any>;
  cancel: () => void;
}

const makeCancellable = (promise: Promise<any>): CancellablePromise => {
    let hasCanceled: boolean = false;

    const wrappedPromise = new Promise((resolve, reject) => {
        return promise ? (
            promise.then(
                (val) => hasCanceled ? reject({isCanceled: true}) : resolve(val),
                (error) => hasCanceled ? reject({isCanceled: true}) : reject(error)
            )
        ) : resolve(promise);
    });

    return {
        promise: wrappedPromise,
        cancel() {
            hasCanceled = true;
        },
    };
};
```

**Comment by richard-ejem** [comment]
@Alex1100 I'm afraid you didn't get my point. This "works" the same as the solution in the referred article - yes, it prevents calling setState on unmounted component, but also it leaks the memory the same way as using `isMounted` method. Javascript engine must keep all callbacks waiting for the promise in the memory until it is either resolved or rejected.

It is usually not a big deal, because you usually don't have so much open promises for so long time to actually be concerned about leaking memory, but the article seems to propose solution to this problem  that is actually not a solution at all - and that was my point.

**Comment by Alex1100** [comment]
@richard-ejem Interesting, yea I would need to test a few things.

**Comment by gaurav5430** [comment]
I get your point and have similar doubts.

Asked here: 
https://twitter.com/dan_abramov/status/1034101342376222723
https://stackoverflow.com/questions/63236205/react-setstate-in-unmounted-component
